Source/Core/Core/PowerPC/JitArm64/Jit.cpp

```diff
@@ -450,20 +462,26 @@ void JitArm64::WriteExit(Arm64Gen::ARM64Reg dest, bool LK, u32 exit_address_afte
   }
 }
 
-void JitArm64::FakeLKExit(u32 exit_address_after_return)
+void JitArm64::FakeLKExit(u32 exit_address_after_return, ARM64Reg exit_address_after_return_reg)
 {
   if (!m_enable_blr_optimization)
     return;
 
   // We may need to fake the BLR stack on inlined CALL instructions.
   // Else we can't return to this location any more.
   gpr.Lock(ARM64Reg::W30);
```

Comment:
There's a potential for a subtle bug here if `exit_address_after_return_reg == ARM64Reg::W30`. The register cache can't correctly handle nested locks an unlocks.

I think the simplest solution would be to skip locking and unlocking W30 if `exit_address_after_return_reg == ARM64Reg::W30`. The clobbering of W30 doesn't happen until we're done using `exit_address_after_return_reg`, so it should be fine. (As long as the calling function doesn't need to use `exit_address_after_return_reg` after this function returns, which currently isn't the case, but still is a little fragile...)

Reply:
Or, alternatively: Instead of calling Lock in this function, make the caller call gpr.Lock(W30) and pass W30 as a parameter to this function. Then use that parameter as an argument for the BL (which is where W30 actually gets used).

Reply:
BL doesn't take register arguments, isn't it always using LR/W30, or am i misunderstanding something?

Reply:
Oh, I thought it did, with X30 just being the default. Alright then, let's go with the other suggestion, as you've already done.

Reply:
Yeah, now that I look at the instruction encoding, it has a whole 26 bits for the offset, so they couldn't fit a register in there. That's sensible.